src/sentry/workflow_engine/migration_helpers/issue_alert_migration.py

```diff
@@ -89,14 +90,22 @@ def _create_detector_lookup(self) -> Detector:
                 error_detector = Detector(type=ErrorGroupType.slug, project=self.project)
 
         else:
-            error_detector, _ = Detector.objects.get_or_create(
-                type=ErrorGroupType.slug,
-                project=self.project,
-                defaults={"config": {}, "name": "Error Detector"},
-            )
-            _, created = AlertRuleDetector.objects.get_or_create(
-                detector=error_detector, rule_id=self.rule.id
+            lock = locks.get(
+                f"workflow-engine-project-error-detector:{self.project.id}",
+                duration=10,
+                name="workflow_engine_issue_alert",
             )
+            with lock.acquire():
+                error_detector, _ = Detector.objects.get_or_create(
+                    type=ErrorGroupType.slug,
+                    project=self.project,
+                    defaults={"config": {}, "name": "Error Detector"},
+                )
+                _, created = AlertRuleDetector.objects.get_or_create(
+                    detector=error_detector, rule_id=self.rule.id
+                )
+
+                # Detector is required for migration, migration should fail if unable to acquire lock
```

Comment:
how will these errors bubble up in the system if we can't acquire the lock? should we have anything defensive to catch those issues?

Reply:
we want it to bubble up so it will rollback the transaction, for example in the API
https://github.com/getsentry/sentry/blob/c697529b9046daa83b33a4134bd0544cf091ff75/src/sentry/projects/project_rules/creator.py#L32-L47

and in the project creation receiver
https://github.com/getsentry/sentry/blob/c697529b9046daa83b33a4134bd0544cf091ff75/src/sentry/receivers/rules.py#L44-L54

the error should rollback the transaction